src/core/CorePool.ts

```diff
@@ -311,11 +343,63 @@ export class CorePool {
     tickUpper: number,
     liquidityDelta: JSBI
   ): { position: Position; amount0: JSBI; amount1: JSBI } {
-    // TODO
+    let amount0: JSBI = ZERO,
+      amount1: JSBI = ZERO;
+
+    let position: Position = this.getPosition(owner, tickLower, tickUpper);
+
+    // use assert
+    if (tickLower > tickUpper) {
+      // error handling
+      console.error("[Error]: tickLower upper than tickUpper");
+    } else if (tickLower < TickMath.MIN_TICK) {
+      console.error("[Error]: tickLower lower than MIN_TICK");
+    } else if (tickUpper > TickMath.MAX_TICK) {
+      console.error("[Error]: tickUpper bigger than MAX_TICK");
+    } else {
+      // check ticks pass, update position
+      this.updatePosition(owner, tickLower, tickUpper, liquidityDelta);
+      // use switch or pattern matching
+      // check if liquidity happen add() or remove()
+      if (JSBI.notEqual(liquidityDelta, ZERO)) {
+        if (this._tickCurrent < tickLower) {
+          amount0 = SqrtPriceMath.getAmount0Delta(
+            TickMath.getSqrtRatioAtTick(tickLower),
+            TickMath.getSqrtRatioAtTick(tickUpper),
+            liquidityDelta
+          );
+        } else if (tickLower < this._tickCurrent) {
+          amount0 = SqrtPriceMath.getAmount0Delta(
+            TickMath.getSqrtRatioAtTick(tickLower),
```

Comment:
The implementation of contract is solid enough for mainnet running as well as audit so let's follow that.
`TickMath.getSqrtRatioAtTick(tickLower)` -> `this.sqrtPriceX96`

Reply:
A little bit confused here, within my understanding, `TickMath.getSqrtRatioAtTick(tickLower/tickUpper)` contains two different input param, is that possible to use `this.sqrtPriceX96` which is a single value to represent two different input params returned value. Also sqrtPriceX96 represent the amount0/amount1 value things, confused here to use this one replace tickLower input and tickUpper input.

Reply:
This has something to do with liquidity model. If current price is within position range, then current price need to be involved with calculation.
For reference: https://liaoph.com/uniswap-v3-2/#%E4%BB%8E-token-%E6%95%B0%E8%AE%A1%E7%AE%97%E6%B5%81%E5%8A%A8%E6%80%A7-l

Reply:
Thx for ref doc, after reading contract code and ref doc, I think I'll correct this part firstly, and then try to figure out how it's handle the sqrtPrice and ticks to get amount0 and amount1, which divided into three part by position.